### DISCUSSION TITLE
How to use client extensions in Nestjs

### DISCUSSION THREAD

**Discussion by weeptao** [prisma/prisma#25790] [Q&A]
### Question

I want to use client extensions in Nestjs, but I don't know how to configure them correctly。
My Prisma version is 5.19

### How to reproduce (optional)

![image](https://github.com/user-attachments/assets/60a3e671-cdb8-4559-8061-627b1c76faa3)

### Expected behavior (optional)

when i run prisma.test.findMany() ,I have obtained two pieces of data。And the log in the findMany method is not output either
![image](https://github.com/user-attachments/assets/3d24b500-db00-4754-90a5-6d5f3338f123)

### Information about Prisma Schema, Client Queries and Environment (optional)

```prisma
// Add your schema.prisma
```

```ts
// Add any relevant Prisma Client queries here
```

- OS: 
- Database: 
- Node.js version: 

- Run `prisma -v` to see your Prisma version and paste it

**Comment by ludralph** [comment]
Hi @weeptao 

In this example, We create a `PrismaService` that extends `PrismaClient`. In the constructor, we use `$extends` to add a custom `signUp` method to the `user` model. We implement `OnModuleInit` to connect to the database when the module initializes.
```js
import { Injectable, OnModuleInit } from '@nestjs/common';
import { PrismaClient } from '@prisma/client';

@Injectable()
export class PrismaService extends PrismaClient implements OnModuleInit {
  constructor() {
    super();
    this.$extends({
      model: {
        user: {
          async signUp(email: string) {
            await this.create({ data: { email } });
          },
        },
      },
    });
  }

  async onModuleInit() {
    await this.$connect();
  }
}
```
You can then use this service in your NestJS controllers or other services like this:
```js
@Injectable()
export class UserService {
  constructor(private prisma: PrismaService) {}

  async createUser(email: string) {
    await this.prisma.user.signUp(email);
  }
}
```

You can also take a look at this [issue](https://github.com/prisma/prisma/issues/18628)

**Reply by tutods** [reply]
@ludralph I'm trying to fill a code with nanoid by default with an extension but isn't working
```ts
import { Injectable, Logger, type OnModuleDestroy, type OnModuleInit } from '@nestjs/common';
import { PrismaClient } from '@prisma/client';
import { customAlphabet } from 'nanoid';
import { createSoftDeleteExtension } from 'prisma-extension-soft-delete';

@Injectable()
export class PrismaService extends PrismaClient implements OnModuleInit, OnModuleDestroy {
  private readonly logger = new Logger(PrismaService.name);
  #nanoid: (size?: number) => string;

  constructor() {
    super();
    const nanoid = customAlphabet('ABCDEFGHIJKLMNOPQRSTUVWXYZ0123456789', 4);

    this.$extends({
      model: {
        customer: {
          async create({ args, query }) {
            const customerCode = nanoid();
            return query({
              ...args,
              data: {
                ...args.data,
                code: customerCode,
              },
            });
          },
        },
      },
    });
  }

  async onModuleInit() {
    // this.logger.log(`CODE: ${this.#nanoid()}`);

    await this.$connect();

    this.$extends(
      createSoftDeleteExtension({
        models: {
          User: true,
          Customer: true,
        },
        defaultConfig: {
          field: 'deletedAt',
          createValue: (deleted) => {
            if (deleted) {
              return new Date();
            }
            return null;
          },
        },
      }),
    );
  }

  async onModuleDestroy() {
    await this.$disconnect();
  }
}

```

**Reply by tutods** [reply]
I try a lot of ways and nothing works and I need to send the `code` as empty to prisma not complaining: 
```ts
return await this.prismaService.customer.create({ data: { ...customer, code: '' } });
```

**Reply by ludralph** [reply]
Hi @tutods 

Can you please create a new discussion?

**Reply by worldzhy** [reply]
> Hi @weeptao
> 
> In this example, We create a `PrismaService` that extends `PrismaClient`. In the constructor, we use `$extends` to add a custom `signUp` method to the `user` model. We implement `OnModuleInit` to connect to the database when the module initializes.
> 
> ```js
> import { Injectable, OnModuleInit } from '@nestjs/common';
> import { PrismaClient } from '@prisma/client';
> 
> @Injectable()
> export class PrismaService extends PrismaClient implements OnModuleInit {
>   constructor() {
>     super();
>     this.$extends({
>       model: {
>         user: {
>           async signUp(email: string) {
>             await this.create({ data: { email } });
>           },
>         },
>       },
>     });
>   }
> 
>   async onModuleInit() {
>     await this.$connect();
>   }
> }
> ```
> 
> You can then use this service in your NestJS controllers or other services like this:
> 
> ```js
> @Injectable()
> export class UserService {
>   constructor(private prisma: PrismaService) {}
> 
>   async createUser(email: string) {
>     await this.prisma.user.signUp(email);
>   }
> }
> ```
> 
> You can also take a look at this [issue](https://github.com/prisma/prisma/issues/18628)

Actually, there is no 'signUp ' under this.prisma.user . Have you ever tried to run the code?

**Comment by ludralph** [comment]
Hi there,

To keep our discussions organized and focused on the most relevant topics, we’re reviewing and tidying up our backlog. As part of this process, we’re closing discussions that haven’t had any recent activity and appear to be outdated.

If this discussion is still important to you or unresolved, we’d love to hear from you! Feel free to reopen it or start a new one with updated details.

For more details about our priorities and vision for the future of Prisma ORM, check out our latest blog post: https://www.prisma.io/blog/prisma-orm-manifesto.

Thank you for your understanding and being part of the community!;

**Comment by ludndev** [comment]
See https://github.com/prisma/prisma/issues/18628#issuecomment-3213927054
